### ISSUE TITLE
fold(unfold(x)) != x

### ISSUE THREAD

**Issue by nadavbh12** [pytorch/pytorch#7519]
## Issue description

Shouldn’t these operations be each other's inverse? If not, How do I inverse this operation?
I tried to run this on MNIST images, and visually I also see a difference at the edges of the digits.

## Code example
```In[1]: x = torch.rand(32, 3, 16, 16)
In[2]: torch.equal(F.fold(F.unfold(x, 3), x.shape[2], 3), x)
Out[2]: False
```

## System Info
```PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.2 LTS
GCC version: (Ubuntu 5.4.1-2ubuntu1~16.04) 5.4.1 20160904
CMake version: version 3.6.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration: 
GPU 0: GeForce GTX TITAN X
GPU 1: GeForce GTX TITAN X

Nvidia driver version: 384.111
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.5.1.10
/usr/lib/libcudnn_static.a
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.7.0.5
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.14.0)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.0           py36_cuda8.0.61_cudnn7.1.2_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
```

**Comment by fmassa** [comment]
Those operations are not supposed to be exactly the inverse of the other.
What happens is that `unfold` repeats elements of one tensor into another, and `fold` sums the corresponding elements back. If instead an average was performed, then you'd obtain the same results.

```python
x = torch.arange(1, 17).reshape(1, 1 ,4, 4)
y = torch.nn.functional.unfold(x, 3, padding=1)
z = torch.nn.functional.fold(y, 4, 3, padding=1)
print(z / x)
```
yields
```
tensor([[[[ 4.,  6.,  6.,  4.],
          [ 6.,  9.,  9.,  6.],
          [ 6.,  9.,  9.,  6.],
          [ 4.,  6.,  6.,  4.]]]])
```
Note that except from the boundaries (because the tensor is zero-padded), the output is 9 times the input (because kernel size is 3).

Also, note that you were incorrectly passing the wrong output size for `fold`, as you were not padding the input image when doing `unfold`, the output size would be smaller.

Closing as this is not a bug.

**Comment by nadavbh12** [comment]
Thanks for the detailed response!
